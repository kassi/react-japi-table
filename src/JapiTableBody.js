import React from 'react';
import PropTypes from 'prop-types';

export default class JapiTableBody extends React.Component {
  constructor (props) {
    super(props);

    this.state = {
      data: [] // stores adapted data taken from props data, enriched with group info
    };
  }

  componentDidMount () {
    this.preProcessData(this.props);
  }

  componentWillReceiveProps (newProps) {
    this.preProcessData(newProps);
  }

  preProcessData (newProps) {
    if (newProps) {
      const rawColumns = newProps.columns ? newProps.columns : this.props.columns;
      const rawData = (newProps.data ? newProps.data.data : this.props.data ? this.props.data.data : []) || [];
      const groupColumn = rawColumns.find((elem, i) => { return elem.group; });
      let data = [];
      if (groupColumn) {
        let groups = {};
        rawData.forEach((elem, i) => {
          const groupKey = groupColumn['key'];
          const groupValue = this.getAttribute(elem, groupKey);
          if (!groups[groupValue]) {
            groups[groupValue] = 1;
            const groupData = {
              group: groupKey,
              column: groupColumn,
              colSpan: groupColumn['groupColumn'] === 'hide' ? rawColumns.length - 1 : rawColumns.length
            };
            data.push(Object.assign({}, elem, groupData));
          }
          data.push(elem);
        });
      } else if (rawData) {
        data = rawData;
      }
      this.setState({data: data});
    }
  }

  getAttribute (data, key) {
    let result = key === 'id' ? data[key]
      : (data.attributes && data.attributes[key]) ? data.attributes[key]
      : (data.relationships && data.relationships[key]) ? this.getIncluded(data.relationships[key].data)
      : undefined;
    return result;
  }

  getIncluded (relation_or_array) {
    let relations = Array.isArray(relation_or_array) ? relation_or_array : [relation_or_array];
    return relations.map((relation, i) => {
      return this.props.data.included.find((elem) => {
        return elem['type'] == relation['type'] && elem['id'] == relation['id'];
      });
    });
  }

  getValueOfKeyPath (keyPath, data) {
    if (keyPath.length == 1) {
      return [this.getAttribute(data, keyPath[0]), data];
    } else {
      const key = keyPath.shift();
      const relation = data.relationships && data.relationships[key] && data.relationships[key]['data'];
      if (relation) {
        const reference = this.getIncluded(relation)[0];
        return this.getValueOfKeyPath(keyPath, reference);
      }
      console.warn('relationship not found in data:', key);
      return undefined;
    }
  }

  getValue (data, column) {
    const key = column.key;
    const keyPath = key.split('.');
    let value_and_object = this.getValueOfKeyPath(keyPath, data, column);
    let value = Array.isArray(value_and_object) ? value_and_object[0] : value_and_object;
    let value_object = Array.isArray(value_and_object) ? value_and_object[1] : data;

    if (column.renderValue) {
      return column.renderValue(value, {rowData: data, objectData: value_object, getIncluded: this.getIncluded.bind(this)});
    }
    if (column.autolink && !!(value_object.links && value_object.links.self)) {
      return (
        <a href={value_object.links.self}>{value}</a>
      );
    }
    return value;
  }

  render () {
    return (
      <tbody>
        {this.state.data.map((row, i) => {
          if (row.group) {
            return (
              <tr key={i} className={row.column.groupRowClassName}>
                <td colSpan={row.colSpan}>
                  {this.getValue(row, row.column)}
                </td>
              </tr>
            );
          } else {
            return (
              <tr key={i}>
                {this.props.columns.map((column, j) => {
                  if (!(column.group && column.groupColumn === 'hide')) {
                    return (
                      <td className={column['cellClassName']} key={'' + i + '-' + j}>
                        {this.getValue(row, column)}
                      </td>
                    );
                  }
                })}
              </tr>
            );
          }
        })}
      </tbody>
    );
  }
}

JapiTableBody.propTypes = {
  columns: PropTypes.array.isRequired,
  data: PropTypes.object.isRequired
};
