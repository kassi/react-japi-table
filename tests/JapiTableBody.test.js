import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import JapiTableBody from '../src/JapiTableBody';

describe('JapiTableHead', () => {
  it('renders a tbody', () => {
    expect(shallow(<JapiTableBody />).type()).toBe('tbody');
  });

  it('renders a table row for each data entry', () => {
    const columns = [
      { key: 'id', header: 'Id' },
      { key: 'title', header: 'Title' }
    ];
    const data = {
      data: [
        {
          id: '1',
          type: '_type'
        },
        {
          id: '2',
          type: '_type'
        }
      ]
    };
    const node = renderer.create(<JapiTableBody columns={columns} data={data} />);
    expect(node).toMatchSnapshot();
  });

  it('renders an optional header cell className', () => {
    const columns = [
      {
        key: 'id',
        header: 'Id',
        cellClassName: 'class1'
      },
      {
        key: 'title',
        header: 'Title',
        cellClassName: 'class2'
      }
    ];
    const data = {
      data: [
        {
          id: '1',
          type: '_type'
        },
        {
          id: '2',
          type: '_type'
        }
      ]
    };
    const node = renderer.create(<JapiTableBody columns={columns} data={data} />);
    expect(node).toMatchSnapshot();
  });

  it('renders appropriate cell values', () => {
    const columns = [
      {
        key: 'id',
        header: 'Id'
      },
      {
        key: 'title',
        header: 'Title'
      }
    ];
    const data = {
      data: [
        {
          id: '101',
          type: '_type',
          attributes: {
            title: 'Title 1'
          }
        },
        {
          id: '102',
          type: '_type',
          attributes: {
            title: 'Title 2'
          }
        }
      ]
    };
    const node = renderer.create(<JapiTableBody columns={columns} data={data} />);
    expect(node).toMatchSnapshot();
  });

  it('renders cell values with render function', () => {
    const columns = [
      {
        key: 'id',
        header: 'Id'
      },
      {
        key: 'date',
        header: 'Date',
        renderValue: function (cellData) {
          return (
            <time datetime={cellData}>{cellData}</time>
          );
        }
      }
    ];
    const data = {
      data: [
        {
          id: '101',
          type: '_type',
          attributes: {
            date: '2017-04-10'
          }
        },
        {
          id: '102',
          type: '_type',
          attributes: {
            date: '2017-04-12'
          }
        }
      ]
    };
    const node = renderer.create(<JapiTableBody columns={columns} data={data} />);
    expect(node).toMatchSnapshot();
  });

  it('renders row values with render function', () => {
    const columns = [
      {
        key: 'id',
        header: 'Id'
      },
      {
        key: 'date',
        header: 'Date',
        renderValue: function (cellData, { rowData }) {
          return (
            <time datetime={cellData}>{rowData.attributes.other}</time>
          );
        }
      }
    ];
    const data = {
      data: [
        {
          id: '101',
          type: '_type',
          attributes: {
            date: '2017-04-10',
            other: 'some other value'
          }
        },
        {
          id: '102',
          type: '_type',
          attributes: {
            date: '2017-04-12',
            other: 'another other value'
          }
        }
      ]
    };
    const node = renderer.create(<JapiTableBody columns={columns} data={data} />);
    expect(node).toMatchSnapshot();
  });

  it('renders cell values of 1:1 relationships', () => {
    const columns = [
      {
        key: 'id',
        header: 'Id'
      },
      {
        key: 'game.title',
        header: 'Titel'
      }
    ];
    const data = {
      data: [
        {
          id: '101',
          type: 'matches',
          attributes: {
            date: '2017-04-10'
          },
          relationships: {
            game: {
              data: {
                id: '201',
                type: 'games'
              }
            }
          }
        },
        {
          id: '102',
          type: 'matches',
          attributes: {
            date: '2017-04-12'
          },
          relationships: {
            game: {
              data: {
                id: '202',
                type: 'games'
              }
            }
          }
        }
      ],
      included: [
        {
          id: '201',
          type: 'games',
          attributes: {
            title: 'Game no 1'
          }
        },
        {
          id: '202',
          type: 'games',
          attributes: {
            title: 'Game no 2'
          }
        }
      ]
    };
    const node = renderer.create(<JapiTableBody columns={columns} data={data} />);
    expect(node).toMatchSnapshot();
  });

  it('renders object values of relationships with render function', () => {
    const columns = [
      {
        key: 'id',
        header: 'Id'
      },
      {
        key: 'game.title',
        header: 'Titel',
        renderValue: function (cellData, { objectData }) {
          return objectData.attributes.other;
        }
      }
    ];
    const data = {
      data: [
        {
          id: '101',
          type: 'matches',
          attributes: {
            date: '2017-04-10'
          },
          relationships: {
            game: {
              data: {
                id: '201',
                type: 'games'
              }
            }
          }
        },
        {
          id: '102',
          type: 'matches',
          attributes: {
            date: '2017-04-12'
          },
          relationships: {
            game: {
              data: {
                id: '202',
                type: 'games'
              }
            }
          }
        }
      ],
      included: [
        {
          id: '201',
          type: 'games',
          attributes: {
            title: 'Game no 1',
            other: 'game 1 other'
          }
        },
        {
          id: '202',
          type: 'games',
          attributes: {
            title: 'Game no 2',
            other: 'game 2 other'
          }
        }
      ]
    };
    const node = renderer.create(<JapiTableBody columns={columns} data={data} />);
    expect(node).toMatchSnapshot();
  });

  it('adds links to unrendered value', () => {
    const columns = [
      {
        key: 'id',
        header: 'Id',
        autolink: true
      },
      {
        key: 'game.title',
        header: 'Titel',
        autolink: true
      }
    ];
    const data = {
      data: [
        {
          id: '101',
          type: 'matches',
          attributes: {
            date: '2017-04-10'
          },
          links: {
            self: 'http://example.com/matches/101'
          },
          relationships: {
            game: {
              data: {
                id: '201',
                type: 'games'
              }
            }
          }
        },
        {
          id: '102',
          type: 'matches',
          attributes: {
            date: '2017-04-12'
          },
          links: {
            self: 'http://example.com/matches/102'
          },
          relationships: {
            game: {
              data: {
                id: '202',
                type: 'games'
              }
            }
          }
        }
      ],
      included: [
        {
          id: '201',
          type: 'games',
          attributes: {
            title: 'Game no 1'
          },
          links: {
            self: 'http://example.com/games/201'
          }
        },
        {
          id: '202',
          type: 'games',
          attributes: {
            title: 'Game no 2'
          },
          links: {
            self: 'http://example.com/games/202'
          }
        }
      ]
    };
    const node = renderer.create(<JapiTableBody columns={columns} data={data} />);
    expect(node).toMatchSnapshot();
  });

  it('has access to related objects via key path and renderValue function', () => {
    const columns = [
      {
        key: 'id',
        header: 'Id'
      },
      {
        key: 'date',
        header: 'Date'
      },
      {
        key: 'participations',
        header: 'Participations',
        renderValue: function (cellData, { rowData }) {
          return cellData.length;
        }
      },
      {
        key: 'participations',
        header: 'Participations',
        renderValue: function (cellData, { objectData, getIncluded }) {
          const winners = cellData.filter((elem, i) => {
            return elem.attributes.winner;
          });
          const names = winners.map((elem, i) => {
            return getIncluded(elem.relationships.player.data).map((player) => {
              return player.attributes.name;
            });
          });
          return names.join(', ');
        }
      }
    ];
    const data = {
      data: [
        {
          id: '201',
          type: 'matches',
          attributes: {
            date: '2017-04-10'
          },
          relationships: {
            participations: {
              data: [
                {
                  id: '301',
                  type: 'participations'
                },
                {
                  id: '302',
                  type: 'participations'
                }
              ]
            }
          }
        }
      ],
      included: [
        {
          id: '301',
          type: 'participations',
          attributes: {
            points: 51,
            winner: true
          },
          relationships: {
            player: {
              data: {
                id: '401',
                type: 'players'
              }
            }
          }
        },
        {
          id: '302',
          type: 'participations',
          attributes: {
            points: 47,
            winner: false
          },
          relationships: {
            player: {
              data: {
                id: '402',
                type: 'players'
              }
            }
          }
        },
        {
          id: '401',
          type: 'players',
          attributes: {
            name: 'player one'
          }
        },
        {
          id: '402',
          type: 'players',
          attributes: {
            name: 'player two'
          }
        }
      ]
    };
    const node = renderer.create(<JapiTableBody columns={columns} data={data} />);
    expect(node).toMatchSnapshot();
  });
});
