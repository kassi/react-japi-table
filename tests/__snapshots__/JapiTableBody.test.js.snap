// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JapiTableBody render adds a row above each data row for a group value 1`] = `
<tbody>
  <tr
    className={undefined}
  >
    <td
      colSpan={2}
    >
      2017-01-03
    </td>
  </tr>
  <tr>
    <td
      className={undefined}
    >
      Title 1
    </td>
    <td
      className={undefined}
    >
      30
    </td>
  </tr>
  <tr>
    <td
      className={undefined}
    >
      Title 2
    </td>
    <td
      className={undefined}
    >
      40
    </td>
  </tr>
  <tr
    className={undefined}
  >
    <td
      colSpan={2}
    >
      2017-01-02
    </td>
  </tr>
  <tr>
    <td
      className={undefined}
    >
      Title 3
    </td>
    <td
      className={undefined}
    >
      50
    </td>
  </tr>
</tbody>
`;

exports[`JapiTableBody render adds links to unrendered value 1`] = `
<tbody>
  <tr>
    <td
      className={undefined}
    >
      <a
        href="http://example.com/matches/101"
      >
        101
      </a>
    </td>
    <td
      className={undefined}
    >
      <a
        href="http://example.com/games/201"
      >
        Game no 1
      </a>
    </td>
  </tr>
  <tr>
    <td
      className={undefined}
    >
      <a
        href="http://example.com/matches/102"
      >
        102
      </a>
    </td>
    <td
      className={undefined}
    >
      <a
        href="http://example.com/games/202"
      >
        Game no 2
      </a>
    </td>
  </tr>
</tbody>
`;

exports[`JapiTableBody render has access to related objects via key path and renderValue function 1`] = `
<tbody>
  <tr>
    <td
      className={undefined}
    >
      201
    </td>
    <td
      className={undefined}
    >
      2017-04-10
    </td>
    <td
      className={undefined}
    >
      2
    </td>
    <td
      className={undefined}
    >
      player one
    </td>
  </tr>
</tbody>
`;

exports[`JapiTableBody render hides a group column when groupColumn is set to "hide" 1`] = `
<tbody>
  <tr
    className={undefined}
  >
    <td
      colSpan={1}
    >
      2017-01-01
    </td>
  </tr>
  <tr>
    <td
      className={undefined}
    >
      Title 1
    </td>
  </tr>
  <tr
    className={undefined}
  >
    <td
      colSpan={1}
    >
      2017-01-02
    </td>
  </tr>
  <tr>
    <td
      className={undefined}
    >
      Title 2
    </td>
  </tr>
</tbody>
`;

exports[`JapiTableBody render renders a table row for each data entry 1`] = `
<tbody>
  <tr>
    <td
      className={undefined}
    >
      1
    </td>
    <td
      className={undefined}
    />
  </tr>
  <tr>
    <td
      className={undefined}
    >
      2
    </td>
    <td
      className={undefined}
    />
  </tr>
</tbody>
`;

exports[`JapiTableBody render renders an optional header cell className 1`] = `
<tbody>
  <tr>
    <td
      className="class1"
    >
      1
    </td>
    <td
      className="class2"
    />
  </tr>
  <tr>
    <td
      className="class1"
    >
      2
    </td>
    <td
      className="class2"
    />
  </tr>
</tbody>
`;

exports[`JapiTableBody render renders appropriate cell values 1`] = `
<tbody>
  <tr>
    <td
      className={undefined}
    >
      101
    </td>
    <td
      className={undefined}
    >
      Title 1
    </td>
  </tr>
  <tr>
    <td
      className={undefined}
    >
      102
    </td>
    <td
      className={undefined}
    >
      Title 2
    </td>
  </tr>
</tbody>
`;

exports[`JapiTableBody render renders cell values of 1:1 relationships 1`] = `
<tbody>
  <tr>
    <td
      className={undefined}
    >
      101
    </td>
    <td
      className={undefined}
    >
      Game no 1
    </td>
  </tr>
  <tr>
    <td
      className={undefined}
    >
      102
    </td>
    <td
      className={undefined}
    >
      Game no 2
    </td>
  </tr>
</tbody>
`;

exports[`JapiTableBody render renders cell values with render function 1`] = `
<tbody>
  <tr>
    <td
      className={undefined}
    >
      101
    </td>
    <td
      className={undefined}
    >
      <time
        datetime="2017-04-10"
      >
        2017-04-10
      </time>
    </td>
  </tr>
  <tr>
    <td
      className={undefined}
    >
      102
    </td>
    <td
      className={undefined}
    >
      <time
        datetime="2017-04-12"
      >
        2017-04-12
      </time>
    </td>
  </tr>
</tbody>
`;

exports[`JapiTableBody render renders object values of relationships with render function 1`] = `
<tbody>
  <tr>
    <td
      className={undefined}
    >
      101
    </td>
    <td
      className={undefined}
    >
      game 1 other
    </td>
  </tr>
  <tr>
    <td
      className={undefined}
    >
      102
    </td>
    <td
      className={undefined}
    >
      game 2 other
    </td>
  </tr>
</tbody>
`;

exports[`JapiTableBody render renders row values with render function 1`] = `
<tbody>
  <tr>
    <td
      className={undefined}
    >
      101
    </td>
    <td
      className={undefined}
    >
      <time
        datetime="2017-04-10"
      >
        some other value
      </time>
    </td>
  </tr>
  <tr>
    <td
      className={undefined}
    >
      102
    </td>
    <td
      className={undefined}
    >
      <time
        datetime="2017-04-12"
      >
        another other value
      </time>
    </td>
  </tr>
</tbody>
`;
